File: packages/sdk-go/retry.go
package rollgate

import (
	"context"
	"math"
	"math/rand"
	"time"
)

// RetryResult holds the result of a retry operation.
type RetryResult struct {
	Success  bool
	Attempts int
	Error    error
}

// Retryer handles retry logic with exponential backoff.
type Retryer struct {
	config RetryConfig
}

// NewRetryer creates a new Retryer with the given config.
func NewRetryer(config RetryConfig) *Retryer {
	return &Retryer{config: config}
}

// Do executes the function with retry logic.
func (r *Retryer) Do(ctx context.Context, fn func() error) RetryResult {
	var lastErr error
	attempts := 0

	for attempts <= r.config.MaxRetries {
		attempts++

		err := fn()
		if err == nil {
			return RetryResult{
				Success:  true,
				Attempts: attempts,
			}
		}

		lastErr = err

		// Don't retry non-retryable errors
		if !IsRetryable(err) {
			return RetryResult{
				Success:  false,
				Attempts: attempts,
				Error:    lastErr,
			}
		}

		// Don't retry if we've exhausted attempts
		if attempts > r.config.MaxRetries {
			break
		}

		// Calculate backoff delay
		delay := r.calculateBackoff(attempts - 1)

		// Wait with context cancellation support
		select {
		case <-ctx.Done():
			return RetryResult{
				Success:  false,
				Attempts: attempts,
				Error:    ctx.Err(),
			}
		case <-time.After(delay):
			// Continue to next attempt
		}
	}

	return RetryResult{
		Success:  false,
		Attempts: attempts,
		Error:    lastErr,
	}
}

// DoWithResult executes the function with retry logic and returns a result.
func (r *Retryer) DoWithResult(ctx context.Context, fn func() (any, error)) (any, RetryResult) {
	var result any
	var lastErr error
	attempts := 0

	for attempts <= r.config.MaxRetries {
		attempts++

		res, err := fn()
		if err == nil {
			return res, RetryResult{
				Success:  true,
				Attempts: attempts,
			}
		}

		lastErr = err

		// Don't retry non-retryable errors
		if !IsRetryable(err) {
			return nil, RetryResult{
				Success:  false,
				Attempts: attempts,
				Error:    lastErr,
			}
		}

		// Don't retry if we've exhausted attempts
		if attempts > r.config.MaxRetries {
			break
		}

		// Calculate backoff delay
		delay := r.calculateBackoff(attempts - 1)

		// Wait with context cancellation support
		select {
		case <-ctx.Done():
			return nil, RetryResult{
				Success:  false,
				Attempts: attempts,
				Error:    ctx.Err(),
			}
		case <-time.After(delay):
			// Continue to next attempt
		}
	}

	return result, RetryResult{
		Success:  false,
		Attempts: attempts,
		Error:    lastErr,
	}
}

// calculateBackoff calculates the delay for the given attempt using exponential backoff.
func (r *Retryer) calculateBackoff(attempt int) time.Duration {
	// Exponential backoff: baseDelay * 2^attempt
	delay := float64(r.config.BaseDelay) * math.Pow(2, float64(attempt))

	// Apply jitter
	if r.config.JitterFactor > 0 {
		jitter := delay * r.config.JitterFactor * (rand.Float64()*2 - 1)
		delay += jitter
	}

	// Ensure delay is not negative
	if delay < 0 {
		delay = 0
	}

	// Cap at max delay
	if delay > float64(r.config.MaxDelay) {
		delay = float64(r.config.MaxDelay)
	}

	return time.Duration(delay)
}

// CalculateBackoff calculates backoff delay for testing/external use.
func CalculateBackoff(attempt int, config RetryConfig) time.Duration {
	r := NewRetryer(config)
	return r.calculateBackoff(attempt)
}

File: packages/sdk-java/src/main/java/io/rollgate/Retry.java
package io.rollgate;

import java.time.Duration;
import java.util.concurrent.Callable;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Retry utilities with exponential backoff.
 */
public class Retry {

    private final Config.RetryConfig config;

    public Retry() {
        this(new Config.RetryConfig());
    }

    public Retry(Config.RetryConfig config) {
        this.config = config;
    }

    public <T> RetryResult<T> execute(Callable<T> action) {
        Exception lastError = null;
        int attempts = 0;

        while (attempts <= config.getMaxRetries()) {
            attempts++;

            try {
                T result = action.call();
                return new RetryResult<>(true, result, null, attempts);
            } catch (Exception e) {
                lastError = e;

                if (!isRetryable(e)) {
                    return new RetryResult<>(false, null, e, attempts);
                }

                if (attempts > config.getMaxRetries()) {
                    break;
                }

                Duration delay = calculateBackoff(attempts - 1);
                try {
                    Thread.sleep(delay.toMillis());
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    return new RetryResult<>(false, null, e, attempts);
                }
            }
        }

        return new RetryResult<>(false, null, lastError, attempts);
    }

    public Duration calculateBackoff(int attempt) {
        long baseMs = config.getBaseDelay().toMillis();
        long delayMs = (long) (baseMs * Math.pow(2, attempt));

        // Apply jitter
        if (config.getJitterFactor() > 0) {
            double jitter = delayMs * config.getJitterFactor();
            delayMs += (long) (ThreadLocalRandom.current().nextDouble(-jitter, jitter));
        }

        // Ensure non-negative
        delayMs = Math.max(0, delayMs);

        // Cap at max delay
        long maxMs = config.getMaxDelay().toMillis();
        delayMs = Math.min(delayMs, maxMs);

        return Duration.ofMillis(delayMs);
    }

    public static boolean isRetryable(Exception e) {
        if (e == null) {
            return false;
        }

        String message = e.getMessage();
        if (message == null) {
            return false;
        }

        String msgLower = message.toLowerCase();

        // Network errors
        String[] networkPatterns = {
            "timeout", "timed out", "connection refused", "connection reset",
            "no route to host", "network is unreachable", "socket"
        };

        for (String pattern : networkPatterns) {
            if (msgLower.contains(pattern)) {
                return true;
            }
        }

        // Server errors (5xx)
        String[] serverPatterns = {"503", "502", "504", "500", "429"};
        for (String pattern : serverPatterns) {
            if (msgLower.contains(pattern)) {
                return true;
            }
        }

        // Rate limiting
        if (msgLower.contains("too many requests")) {
            return true;
        }

        return false;
    }

    /**
     * Result of a retry operation.
     */
    public static class RetryResult<T> {
        private final boolean success;
        private final T data;
        private final Exception error;
        private final int attempts;

        public RetryResult(boolean success, T data, Exception error, int attempts) {
            this.success = success;
            this.data = data;
            this.error = error;
            this.attempts = attempts;
        }

        public boolean isSuccess() {
            return success;
        }

        public T getData() {
            return data;
        }

        public Exception getError() {
            return error;
        }

        public int getAttempts() {
            return attempts;
        }
    }
}
